RFCs/FS-1073-record-constructors.md

```diff
@@ -0,0 +1,63 @@
+# F# RFC FS-1073 - Allow access to Record type constructors
+
+The design suggestion [Allow access to Record type constructors](https://github.com/fsharp/fslang-suggestions/issues/722) has been marked "approved in principle". This RFC covers the detailed proposal.
+
+# Summary
+
+A record type created in F#, such as
+```fsharp
+type PensionData =
+    { Name:string; ProbableNumberOfYearsUntilRetirement:int }
+```
+
+generates a constructor which can be used in C#: 
+
+```csharp
+PensionData r = PensionData("Adam",10)
+```
+
+Using this constructor is not currently allowed in F#:
+```fsharp
+let r = PensionData("Adam",10) // FS0039 The value or constructor 'PensionData' is not defined.
+```
+
+This RFC proposes that this constructor should be accessible from F# code.
+
+# Motivation
+
+Records can of course be created in F# currently, using special F# syntax:
+
+```fsharp
+let r = { Name = "Adam"; ProbableNumberOfYearsUntilRetirement = 10 }
+```
+
+This has a few disadvantages:
+- Verbose syntax compared to using the constructor, with all field names needing to be written out.
+- Lack of explicit specification of the name of the type. Without the binding to `r` plus intellisense it would be hard to determine the type of the record.
+- In order to enter the record, you have to know the field names, and intellisense is not as helpful as it is for constructors.
+- The type inference system around record creation makes guesses based on field names and is arguably not as reliable as when the type is specified.
+
+Therefore it would be good to have, **in addition to** the existing F#-specific syntax, access to the standard constructor for the record type.
+
+# Detailed design
+
+See summary.
+
+# Drawbacks
+
+Organizations preferring using the existing syntax and wanting to enforce this will need to do so through a style guide.
+They may wish to do so because:
+- While the type is less clear, the field names are more clear than with constructors, when looking at code without intellisense.
+- They often like to change the order of fields in a record, which is safer using the existing syntax if the field share the same type.
+
+# Alternatives
+
+Do nothing.
+
+# Compatibility
+
+This is not a breaking change.
+
+# Unresolved questions
+
+None.
```

Comment:
Please add the following unresolved questions

* Should this form be usable in pattern matching?

* Should named arguments be accepted?  Is there a way to make the use of named arguments **mandatory** so that changing field order will not break client code?

* There is a potential future interaction with optional record fields.  In many ways, this proposal makes optional fields in records much more palatable since the existing optional argument syntax can be used.  We should just note the potential interaction and give a link to the suggestion for optional entries in records

* Can the record names be used as first class function values?

* There is a potential for a breaking change for this kind of code:

```fsharp
type R = { a: int }
let R x = { a = x }
```

In this case the second `R` should take precedence over the first `R`

* There is also a potential for a (less serious?) breaking change when a module containing such a `R` is opened out-scoping another `R` definition, e.g.


```fsharp
module M1 = 
    let R x = { a = x }

module M2 = 
    let Q x = x + 1
    type R = { a: int }

open M1
open M2

R 3
```
Note the second `open M2` might now bring a different `R` into scope.  We would have to find a resolution for this I think as it is definitely possible that it would affect existing code.  I suppose the resolution would be to prefer **any** other `R` over the type name.

Reply:
> Should this form be usable in pattern matching?

This makes it a larger request but I'll include that.

> Should named arguments be accepted?

Yes. Added a section making it clear that this is already the case.

> Is there a way to make the use of named arguments mandatory so that changing field order will not break client code?

This sounds like a larger .Net suggestion. It's actually less relevant to records than other types because records have the standard F# record syntax already enables this.

> There is a potential future interaction with optional record fields...

Done

> Can the record names be used as first class function values?

Covered by [a separate issue](https://github.com/fsharp/fslang-suggestions/issues/506).

> There is a potential for a breaking change

Added

Reply:
> Is there a way to make the use of named arguments mandatory so that changing field order will not break client code?

There is a suggestion for this: 
https://github.com/fsharp/fslang-suggestions/issues/414

Reply:
Thanks!